### DISCUSSION TITLE
Tabs don't separate sets of buffers in bufferline

### DISCUSSION THREAD

**Discussion by kapral18** [LazyVim/LazyVim#1657] [Q&A] [answered]
When I create tab 1 with file 1 and then I create tab 2 with file 2, while in Tab1 I can see file 2 in the bufferline. Is this how it's supposed to be? I thought each tab is supposed to isolate its own set of buffers

My config: https://github.com/kapral18/config.nvim/blob/75f286e6a99344820b2dc3c445e7268cfc2b8441/init.lua

![CleanShot 2023-10-10 at 23 23 32@2x](https://github.com/LazyVim/LazyVim/assets/1625373/03171e75-6e0e-411f-bfff-d2635c569a75)
]
![CleanShot 2023-10-10 at 23 23 43@2x](https://github.com/LazyVim/LazyVim/assets/1625373/6651d3ce-52dd-4ba1-b89e-299bfe7348c3)

**Comment by folke** [comment, marked as answer]
Nope, thats how it's supposed to be.
Tabs have all buffers always. The difference is windows. Each tab has its own set of windows

**Reply by dpetka2001** [reply]
Check out [scope.nvim](https://github.com/tiagovla/scope.nvim)

**Reply by kapral18** [reply]
@folke I see :(

@dpetka2001 I applied it but no effect at all

**Reply by kapral18** [reply]
actually managed to make it work by manually calling require("scope").setup({}) in init function, I thought lazy was by default calling pluginName.setup(options) in every plugin that has setup, apparently not in case of this plugin

**Reply by folke** [reply]
you need to add `opts={}`, dont call setup in init.

**Reply by folke** [reply]
`{ "tiagovla/scope.nvim", opts = {} }`

**Reply by dpetka2001** [reply]
```lua
  { 
     "tiagovla/scope.nvim", 
     -- event = { "BufReadPre", "BufNewFile" }, 
     event = "LazyFile", 
     opts = { 
       restore_state = false, 
     }, 
   },
```
This is how i have it set up.

**Reply by kapral18** [reply]
@folke is there any generalized recommendation with regards to how to detect when I need to pass opts ={} and when it's done automatically?

**Reply by dpetka2001** [reply]
You always need to pass `opts = {}` or `config = true` for the plugin's `setup` function to be called by the default `config` function. If you `require(plugin).setup` in the config function, then you must additionaly make sure to pass `opts` to the `config` function, because you're overriding the default `config` function and you have to do the necessary setup on your own. `lazy.nvim`'s docs will help you with this stuff.

**Reply by kapral18** [reply]
Im confused.

In lazy.nvim docs it's clearly stated that plugins can be used just with a string like { "folke/which-key" } looking at the source code of which-key plugin there is a setup() function that can accept nil `opts` value and defaults to `{}` if that happens, and looking at the `lazy.nvim` docs  at the `config` section the default `require(MAIN).setup(opts)` is always called, so I assumed that happens by default, even if `opts` is not mentioned then `require(MAIN).setup(nil)` will be called which should be handled by plugins and most plugins default to `{}` when their setup is called with no `opts`. 

I also checked the scope.nvim source and I see that they default to `{}` as well there.

Where does my reasoning go wrong?

**Reply by folke** [reply]
No, read the docs again. @dpetka2001 is correct.

A plugin's setup is **only** called when you specify `opts` or set `config=true`.

That will use the default `config` implementation.

If you override `config` you need to call setup yourself in that function.

Specifying `{"folke/which-key"}` is perfectly valid, but will **not** call it's setup function automatically.

Not all plugins like which-key require an explicit setup call.

**Reply by kapral18** [reply]
@folke thanks for clearing it up for me, now it's clear. But with all due respect docs don't make it clear even after rereading it many times.

Something like this could be much more structurally clear for me to reason about:

opts | table or fun(LazyPlugin, opts:table) | Default: nil. If present should be a table (will be merged with parent specs), a function that returns a table (replaces parent specs) or function that changes a table. If present, triggers default implementation of `config` option below.
-- | -- | --
config | fun(LazyPlugin, opts:table) or true | Default: nil. If config is `true` it will automatically run require(MAIN).setup(opts). See `opts` option above. Lazy uses several heuristics to determine the plugin's MAIN module automatically based on the plugin's name. If present, is executed when the plugin loads.

**Reply by dpetka2001** [reply]
> The table will be passed to the Plugin.config() function. Setting this value will imply Plugin.config()

It distinctly mentions that setting this value implies `Plugin.config()`.
